File: docs/integrations/tests/logger_test.py
"""
Unit tests for Python logger (logger.py)

Tests buffer management, threading, batch sending, retry logic, and cleanup.
"""

import unittest
import json
import time
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path
import sys
import os

# Add parent directory to path to import logger
sys.path.insert(0, str(Path(__file__).parent.parent / 'python'))

# Load test configuration
test_config_path = Path(__file__).parent / '.test-config.json'
with open(test_config_path) as f:
    TEST_CONFIG = json.load(f)

# Mock HTTP server for testing
received_requests = []
mock_server = None
mock_server_thread = None


class MockHTTPHandler(BaseHTTPRequestHandler):
    """Mock HTTP handler to capture requests"""
    
    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        body = self.rfile.read(content_length).decode('utf-8')
        
        request_data = {
            'method': 'POST',
            'path': self.path,
            'headers': dict(self.headers),
            'body': json.loads(body) if body else None
        }
        received_requests.append(request_data)
        
        # Validate API key
        if self.headers.get('X-Api-Key') != TEST_CONFIG['apiKey']:
            self.send_response(401)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Invalid API key'}).encode())
        else:
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            response = {
                'success': True,
                'received': len(request_data['body'].get('logs', []))
            }
            self.wfile.write(json.dumps(response).encode())
    
    def log_message(self, format, *args):
        # Suppress HTTP log messages
        pass


def start_mock_server(port=8998):
    """Start mock HTTP server in background thread"""
    global mock_server, mock_server_thread
    
    mock_server = HTTPServer(('localhost', port), MockHTTPHandler)
    mock_server_thread = threading.Thread(target=mock_server.serve_forever, daemon=True)
    mock_server_thread.start()
    time.sleep(0.5)  # Give server time to start


def stop_mock_server():
    """Stop mock HTTP server"""
    global mock_server
    if mock_server:
        mock_server.shutdown()
        mock_server.server_close()


class TestPythonLogger(unittest.TestCase):
    """Test suite for Python logger"""
    
    @classmethod
    def setUpClass(cls):
        """Set up test environment once"""
        start_mock_server()
        
        # Import logger after mock server is ready
        from logger import DevSmithLogger
        cls.Logger = DevSmithLogger
    
    @classmethod
    def tearDownClass(cls):
        """Clean up after all tests"""
        stop_mock_server()
    
    def setUp(self):
        """Reset before each test"""
        global received_requests
        received_requests = []
    
    def test_initialization_valid(self):
        """Test logger initialization with valid config"""
        logger = self.Logger(
            api_key=TEST_CONFIG['apiKey'],
            api_url='http://localhost:8998',
            project_slug=TEST_CONFIG['projectSlug'],
            service_name='test-service'
        )
        
        self.assertIsNotNone(logger)
        self.assertEqual(logger.project_slug, TEST_CONFIG['projectSlug'])
        self.assertEqual(logger.service_name, 'test-service')
    
    def test_initialization_missing_api_key(self):
        """Test logger initialization fails without API key"""
        with self.assertRaises(ValueError) as context:
            self.Logger(
                api_key=None,
                api_url='http://localhost:8998',
                project_slug='test',
                service_name='test'
            )
        self.assertIn('API key is required', str(context.exception))
    
    def test_initialization_missing_project_slug(self):
        """Test logger initialization fails without project slug"""
        with self.assertRaises(ValueError) as context:
            self.Logger(
                api_key='key',
                api_url='http://localhost:8998',
                project_slug=None,
                service_name='test'
            )
        self.assertIn('Project slug is required', str(context.exception))
    
    def test_buffer_management(self):
        """Test logs are added to buffer"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test'
        )
        
        logger.info('Message 1')
        logger.info('Message 2')
        
        self.assertEqual(len(logger.buffer), 2)
    
    def test_buffer_size_limit(self):
        """Test buffer respects custom size limit"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test',
            buffer_size=5
        )
        
        for i in range(10):
            logger.info(f'Message {i}')
        
        time.sleep(1)  # Wait for flush
        self.assertLess(len(logger.buffer), 10)
    
    def test_buffer_flush_on_full(self):
        """Test buffer flushes when full"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test',
            buffer_size=3
        )
        
        logger.info('Message 1')
        logger.info('Message 2')
        logger.info('Message 3')  # Should trigger flush
        
        time.sleep(1)  # Wait for background thread
        self.assertEqual(len(received_requests), 1)
        self.assertEqual(len(received_requests[0]['body']['logs']), 3)
    
    def test_log_levels(self):
        """Test all log levels"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test'
        )
        
        logger.debug('Debug message')
        logger.info('Info message')
        logger.warn('Warning message')
        logger.error('Error message')
        
        self.assertEqual(len(logger.buffer), 4)
        self.assertEqual(logger.buffer[0]['level'], 'DEBUG')
        self.assertEqual(logger.buffer[1]['level'], 'INFO')
        self.assertEqual(logger.buffer[2]['level'], 'WARN')
        self.assertEqual(logger.buffer[3]['level'], 'ERROR')
    
    def test_context_and_tags(self):
        """Test context and tags are included"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test'
        )
        
        logger.info('Message', context={'user_id': 123}, tags=['auth', 'user'])
        
        log_entry = logger.buffer[0]
        self.assertEqual(log_entry['context'], {'user_id': 123})
        self.assertEqual(log_entry['tags'], ['auth', 'user'])
    
    def test_batch_format(self):
        """Test batch request format is correct"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test-service',
            buffer_size=2
        )
        
        logger.info('Message 1')
        logger.info('Message 2')
        
        time.sleep(1)  # Wait for flush
        
        self.assertEqual(len(received_requests), 1)
        request = received_requests[0]
        
        self.assertEqual(request['method'], 'POST')
        self.assertEqual(request['headers']['X-Api-Key'], TEST_CONFIG['apiKey'])
        self.assertEqual(request['body']['project_slug'], TEST_CONFIG['projectSlug'])
        self.assertEqual(len(request['body']['logs']), 2)
    
    def test_batch_required_fields(self):
        """Test batch includes all required fields"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test-service',
            buffer_size=1
        )
        
        logger.info('Test message', context={'key': 'value'}, tags=['tag1'])
        
        time.sleep(1)  # Wait for flush
        
        log_entry = received_requests[0]['body']['logs'][0]
        
        self.assertIn('timestamp', log_entry)
        self.assertEqual(log_entry['level'], 'INFO')
        self.assertEqual(log_entry['message'], 'Test message')
        self.assertEqual(log_entry['service'], 'test-service')
        self.assertEqual(log_entry['context'], {'key': 'value'})
        self.assertEqual(log_entry['tags'], ['tag1'])
    
    def test_time_based_flush(self):
        """Test buffer flushes after interval"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test',
            flush_interval=2  # 2 seconds
        )
        
        logger.info('Message 1')
        
        # Should not have sent yet
        self.assertEqual(len(received_requests), 0)
        
        # Wait for flush interval
        time.sleep(2.5)
        self.assertEqual(len(received_requests), 1)
    
    def test_cleanup_on_close(self):
        """Test buffer flushes on close"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test'
        )
        
        logger.info('Message 1')
        logger.info('Message 2')
        
        logger.close()
        
        time.sleep(0.5)  # Wait for flush
        self.assertEqual(len(received_requests), 1)
        self.assertEqual(len(received_requests[0]['body']['logs']), 2)
    
    def test_threading_safety(self):
        """Test logger is thread-safe"""
        logger = self.Logger(
            TEST_CONFIG['apiKey'],
            'http://localhost:8998',
            TEST_CONFIG['projectSlug'],
            'test',
            buffer_size=100
        )
        
        def log_messages(thread_id, count):
            for i in range(count):
                logger.info(f'Thread {thread_id} - Message {i}')
        
        threads = []
        for i in range(5):
            t = threading.Thread(target=log_messages, args=(i, 10))
            threads.append(t)
            t.start()
        
        for t in threads:
            t.join()
        
        # Should have 50 logs total (5 threads x 10 messages)
        logger.close()
        time.sleep(1)
        
        total_logs = sum(len(req['body']['logs']) for req in received_requests)
        self.assertEqual(total_logs, 50)


if __name__ == '__main__':
    unittest.main(verbosity=2)

File: docs/integration-examples/tests/logger.test.js
/**
 * Unit tests for JavaScript logger (logger.js)
 * 
 * Tests buffer management, batch sending, retry logic, and cleanup.
 */

const fs = require('fs');
const path = require('path');
const assert = require('assert');
const http = require('http');

// Load test configuration
const testConfigPath = path.join(__dirname, '.test-config.json');
const testConfig = JSON.parse(fs.readFileSync(testConfigPath, 'utf8'));

// Mock server for testing
let mockServer;
let receivedRequests = [];

// Start mock server before tests
function startMockServer(port = 8999) {
    return new Promise((resolve) => {
        mockServer = http.createServer((req, res) => {
            let body = '';
            req.on('data', chunk => body += chunk);
            req.on('end', () => {
                const request = {
                    method: req.method,
                    url: req.url,
                    headers: req.headers,
                    body: body ? JSON.parse(body) : null
                };
                receivedRequests.push(request);
                
                // Simulate API responses
                if (req.headers['x-api-key'] !== testConfig.apiKey) {
                    res.writeHead(401, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ error: 'Invalid API key' }));
                } else {
                    res.writeHead(200, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ success: true, received: request.body?.logs?.length || 0 }));
                }
            });
        });
        mockServer.listen(port, () => {
            console.log(`Mock server started on port ${port}`);
            resolve();
        });
    });
}

// Stop mock server after tests
function stopMockServer() {
    return new Promise((resolve) => {
        if (mockServer) {
            mockServer.close(() => {
                console.log('Mock server stopped');
                resolve();
            });
        } else {
            resolve();
        }
    });
}

// Import logger after mock server is ready
let Logger;

describe('JavaScript Logger Unit Tests', function() {
    this.timeout(10000);

    before(async () => {
        await startMockServer();
        
        // Dynamically import logger
        const loggerPath = path.join(__dirname, '../javascript/logger.js');
        const loggerCode = fs.readFileSync(loggerPath, 'utf8');
        
        // Create a module context
        const moduleExports = {};
        const moduleFunc = new Function('module', 'exports', loggerCode);
        moduleFunc({ exports: moduleExports }, moduleExports);
        Logger = moduleExports.DevSmithLogger || moduleExports;
    });

    after(async () => {
        await stopMockServer();
    });

    beforeEach(() => {
        receivedRequests = [];
    });

    describe('Initialization', () => {
        it('should create logger with valid configuration', () => {
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test-service'
            );
            
            assert.ok(logger);
            assert.strictEqual(logger.projectSlug, testConfig.projectSlug);
            assert.strictEqual(logger.serviceName, 'test-service');
        });

        it('should throw error with missing API key', () => {
            assert.throws(() => {
                new Logger(null, 'http://localhost:8999', 'test', 'service');
            }, /API key is required/);
        });

        it('should throw error with missing project slug', () => {
            assert.throws(() => {
                new Logger('key', 'http://localhost:8999', null, 'service');
            }, /Project slug is required/);
        });
    });

    describe('Buffer Management', () => {
        it('should add logs to buffer', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            
            logger.info('Test message 1');
            logger.info('Test message 2');
            
            assert.strictEqual(logger.buffer.length, 2);
        });

        it('should respect custom buffer size', () => {
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test',
                { bufferSize: 5 }
            );
            
            for (let i = 0; i < 10; i++) {
                logger.info(`Message ${i}`);
            }
            
            // Buffer should be cleared after reaching size
            assert.ok(logger.buffer.length < 10);
        });

        it('should trigger flush when buffer is full', (done) => {
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test',
                { bufferSize: 3 }
            );
            
            logger.info('Message 1');
            logger.info('Message 2');
            logger.info('Message 3'); // Should trigger flush
            
            setTimeout(() => {
                assert.strictEqual(receivedRequests.length, 1);
                assert.strictEqual(receivedRequests[0].body.logs.length, 3);
                done();
            }, 500);
        });
    });

    describe('Log Levels', () => {
        it('should log DEBUG level', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            logger.debug('Debug message', { detail: 'value' });
            
            assert.strictEqual(logger.buffer.length, 1);
            assert.strictEqual(logger.buffer[0].level, 'DEBUG');
            assert.strictEqual(logger.buffer[0].message, 'Debug message');
        });

        it('should log INFO level', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            logger.info('Info message');
            
            assert.strictEqual(logger.buffer[0].level, 'INFO');
        });

        it('should log WARN level', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            logger.warn('Warning message');
            
            assert.strictEqual(logger.buffer[0].level, 'WARN');
        });

        it('should log ERROR level', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            logger.error('Error message', { error: 'details' });
            
            assert.strictEqual(logger.buffer[0].level, 'ERROR');
        });
    });

    describe('Context and Tags', () => {
        it('should include context in log entry', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            logger.info('Message', { userId: 123, action: 'login' });
            
            assert.deepStrictEqual(logger.buffer[0].context, { userId: 123, action: 'login' });
        });

        it('should include tags in log entry', () => {
            const logger = new Logger(testConfig.apiKey, 'http://localhost:8999', testConfig.projectSlug, 'test');
            logger.info('Message', {}, ['auth', 'user']);
            
            assert.deepStrictEqual(logger.buffer[0].tags, ['auth', 'user']);
        });
    });

    describe('Batch Sending', () => {
        it('should send batch with correct format', (done) => {
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test',
                { bufferSize: 2 }
            );
            
            logger.info('Message 1');
            logger.info('Message 2');
            
            setTimeout(() => {
                assert.strictEqual(receivedRequests.length, 1);
                const request = receivedRequests[0];
                
                assert.strictEqual(request.method, 'POST');
                assert.strictEqual(request.headers['x-api-key'], testConfig.apiKey);
                assert.strictEqual(request.body.project_slug, testConfig.projectSlug);
                assert.strictEqual(request.body.logs.length, 2);
                
                done();
            }, 500);
        });

        it('should include all required fields in batch', (done) => {
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test-service',
                { bufferSize: 1 }
            );
            
            logger.info('Test message', { key: 'value' }, ['tag1']);
            
            setTimeout(() => {
                const log = receivedRequests[0].body.logs[0];
                
                assert.ok(log.timestamp);
                assert.strictEqual(log.level, 'INFO');
                assert.strictEqual(log.message, 'Test message');
                assert.strictEqual(log.service, 'test-service');
                assert.deepStrictEqual(log.context, { key: 'value' });
                assert.deepStrictEqual(log.tags, ['tag1']);
                
                done();
            }, 500);
        });
    });

    describe('Time-based Flush', () => {
        it('should flush after flush interval', function(done) {
            this.timeout(7000);
            
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test',
                { flushInterval: 2000 } // 2 seconds
            );
            
            logger.info('Message 1');
            
            // Should not have sent yet
            assert.strictEqual(receivedRequests.length, 0);
            
            // Wait for flush interval
            setTimeout(() => {
                assert.strictEqual(receivedRequests.length, 1);
                done();
            }, 2500);
        });
    });

    describe('Retry Logic', () => {
        it('should retry on network failure', function(done) {
            this.timeout(5000);
            
            // Create logger with invalid URL
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:9999', // Non-existent port
                testConfig.projectSlug,
                'test',
                { bufferSize: 1, flushInterval: 1000 }
            );
            
            logger.info('Test message');
            
            // After failed attempt, logs should remain in buffer
            setTimeout(() => {
                assert.ok(logger.buffer.length > 0, 'Logs should be retained after failure');
                done();
            }, 1500);
        });
    });

    describe('Cleanup', () => {
        it('should flush buffer on close', (done) => {
            const logger = new Logger(
                testConfig.apiKey,
                'http://localhost:8999',
                testConfig.projectSlug,
                'test'
            );
            
            logger.info('Message 1');
            logger.info('Message 2');
            
            logger.close();
            
            setTimeout(() => {
                assert.strictEqual(receivedRequests.length, 1);
                assert.strictEqual(receivedRequests[0].body.logs.length, 2);
                done();
            }, 500);
        });
    });
});

// Run tests if executed directly
if (require.main === module) {
    const Mocha = require('mocha');
    const mocha = new Mocha();
    mocha.suite.emit('pre-require', global, null, mocha);
    
    // Add this file
    mocha.addFile(__filename);
    
    // Run tests
    mocha.run((failures) => {
        process.exitCode = failures ? 1 : 0;
    });
}
